source/s3_auto_ranged_put.c

```diff
@@ -942,7 +954,12 @@ struct aws_future_http_message *s_s3_prepare_upload_part(struct aws_s3_request *
         /* Read the body */
         uint64_t offset = 0;
         size_t request_body_size = s_compute_request_body_size(meta_request, request->part_number, &offset);
-        aws_byte_buf_init(&request->request_body, meta_request->allocator, request_body_size);
+        if (request->request_body.capacity == 0) {
+            AWS_FATAL_ASSERT(request->ticket);
+            request->request_body =
+                aws_s3_buffer_pool_acquire_buffer(request->meta_request->client->buffer_pool, request->ticket);
+            request->request_body.capacity = request_body_size;
```

Comment:
is there any chance that `request_body_size` is larger than `meta_request->part_size`? because `meta_request->part_size` is what we used when reserving the ticket

Reply:
no. request_body_size is always smaller than meta_request->part_size. 
request_body_size is either meta_request->part_size or in case of the last part it will be remainder of data. 

ideally we should be able to use bigger buffers than needed in upload, but we currently rely on capacity for some checks in code. i fixed couple places to be more generic, but at this point i dont have complete confidence that it will work correctly with full buffer everywhere. hence why its doing the trick of setting smaller capacity.

will put a comment about it